From 3483cebb83c67c08ad49abdce31a4752d135758f Mon Sep 17 00:00:00 2001
From: John Howard <howardjohn@google.com>
Date: Thu, 24 Mar 2022 14:57:01 -0700
Subject: fuzz: validate Proxy before calling functions (#38092)

Our code makes an assumption this list is >0 and validates that in one
place rather than every callsite
---
 tests/fuzz/networking_core_v1alpha3_fuzzer.go | 21 ++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/tests/fuzz/networking_core_v1alpha3_fuzzer.go b/tests/fuzz/networking_core_v1alpha3_fuzzer.go
index 2ec0804368..81c6521d69 100644
--- a/tests/fuzz/networking_core_v1alpha3_fuzzer.go
+++ b/tests/fuzz/networking_core_v1alpha3_fuzzer.go
@@ -24,8 +24,8 @@
 	"testing"
 
 	fuzz "github.com/AdaLogics/go-fuzz-headers"
-
 	route "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+
 	"istio.io/istio/pilot/pkg/model"
 	"istio.io/istio/pilot/pkg/networking/plugin"
 	"istio.io/istio/tests/fuzz/utils"
@@ -70,6 +70,9 @@ func InternalFuzzbuildGatewayListeners(data []byte) int {
 	if err != nil {
 		return 0
 	}
+	if !proxyValid(proxy) {
+		return 0
+	}
 	lb := &ListenerBuilder{}
 	err = f.GenerateStruct(lb)
 	if err != nil {
@@ -98,6 +101,9 @@ func InternalFuzzbuildSidecarOutboundHTTPRouteConfig(data []byte) int {
 	if err != nil {
 		return 0
 	}
+	if !proxyValid(proxy) {
+		return 0
+	}
 	req := &model.PushRequest{}
 	err = f.GenerateStruct(req)
 	if err != nil {
@@ -120,6 +126,9 @@ func InternalFuzzbuildSidecarInboundListeners(data []byte) int {
 	if err != nil {
 		return 0
 	}
+	if !proxyValid(proxy) {
+		return 0
+	}
 
 	// create fuzzed plugins
 	number, err := f.GetInt()
@@ -196,9 +205,19 @@ func InternalFuzzbuildSidecarOutboundListeners(data []byte) int {
 	if err != nil {
 		return 0
 	}
+	if !proxyValid(proxy) {
+		return 0
+	}
 	cg := NewConfigGenTest(t, to)
 	p := cg.SetupProxy(proxy)
 	listeners := cg.ConfigGen.buildSidecarOutboundListeners(p, cg.env.PushContext)
 	_ = listeners
 	return 1
 }
+
+func proxyValid(p *model.Proxy) bool {
+	if len(p.IPAddresses) == 0 {
+		return false
+	}
+	return true
+}
-- 
2.35.3

